Marlin/stepper.cpp

```diff
@@ -832,6 +868,137 @@ void Stepper::isr() {
+
+    // Start of segment? Check directions.
+    if (segment_steps == 0) {
+      unsigned char dm = last_direction_bits;
+
+      //check directions using dX, dY, etc.
+      #define UPDATE_DIR(AXIS) \
+  	    if (d## AXIS == 7) {} \
+        else if (d## AXIS < 7) SBI(dm, _AXIS(AXIS)); \
+        else CBI(dm, _AXIS(AXIS));
+
+      #if HAS_X_STEP
+        UPDATE_DIR(X);
+      #endif
+      #if HAS_Y_STEP
+        UPDATE_DIR(Y);
+      #endif
+      #if HAS_Z_STEP
+        UPDATE_DIR(Z);
+      #endif
+
+      UPDATE_DIR(E);
+
+      if (dm != last_direction_bits) {
+        last_direction_bits = dm;
+        set_directions();
+      }
+    }
+
+    #define PULSE_STARTC(AXIS) \
+      if (step## AXIS) { _APPLY_STEP(AXIS)(!_INVERT_STEP_PIN(AXIS), 0); }
+
+    // Stop an active pulse, update the position
+    #define PULSE_STOPC(AXIS) \
+      if (step## AXIS) { \
+        count_position[_AXIS(AXIS)] += count_direction[_AXIS(AXIS)]; \
+        _APPLY_STEP(AXIS)(_INVERT_STEP_PIN(AXIS), 0); \
+      }
+
+    // If a minimum pulse time was specified get the timer 0 value
+    // which increments every 4µs on 16MHz and every 3.2µs on 20MHz.
+    // Two or 3 counts of TCNT0 should be a sufficient delay.
+    #if EXTRA_CYCLES_XYZE > 20
+      uint32_t pulse_start = TCNT0;
+    #endif
+
+    #if HAS_X_STEP
+      PULSE_STARTC(X);
+    #endif
+    #if HAS_Y_STEP
+      PULSE_STARTC(Y);
+    #endif
+    #if HAS_Z_STEP
+      PULSE_STARTC(Z);
+    #endif
+
+    PULSE_STARTC(E);
+
+    // For minimum pulse time wait before stopping pulses
+    #if EXTRA_CYCLES_XYZE > 20
+      while (EXTRA_CYCLES_XYZE > (uint32_t)(TCNT0 - pulse_start) * (INT0_PRESCALER)) { /* nada */ }
+      pulse_start = TCNT0;
+    #elif EXTRA_CYCLES_XYZE > 0
+      DELAY_NOPS(EXTRA_CYCLES_XYZE);
+    #endif
+
+    #if HAS_X_STEP
+      PULSE_STOPC(X);
+    #endif
+    #if HAS_Y_STEP
+      PULSE_STOPC(Y);
+    #endif
+    #if HAS_Z_STEP
+      PULSE_STOPC(Z);
+    #endif
+
+    #if DISABLED(ADVANCE) && DISABLED(LIN_ADVANCE)
+      #if ENABLED(MIXING_EXTRUDER)
+        // Always step the single E axis
+        if (counter_E > 0) {
+          counter_E -= current_block->step_event_count;
+          count_position[E_AXIS] += count_direction[E_AXIS];
+        }
+        MIXING_STEPPERS_LOOP(j) {
+          if (counter_m[j] > 0) {
+            counter_m[j] -= current_block->mix_event_count[j];
+            En_STEP_WRITE(j, INVERT_E_STEP_PIN);
+          }
+        }
+      #else // !MIXING_EXTRUDER
+        PULSE_STOPC(E);
+      #endif
+    #endif // !ADVANCE && !LIN_ADVANCE
+
+    //if last step of this chunk (manual modulo)
+    if((step_events_completed16 & 0x3FF) == 0x3FF) {
+      //release chunk so it can be filled again
+      chunk_response[current_block->chunk_idx] = CHUNK_RESPONSE_NONE;
+
+      //keep iterating until we eventually reach our total step_events_completed
+      current_block->chunk_idx = (current_block->chunk_idx + 1) % (NUM_CHUNK_BUFFERS - 1);
```

Comment:
hmm i think i confused a modulo and a bitwise &... not sure why im subtracting one here.